core-models/src/main/scala/com/mesosphere/usi/core/models/RunTemplateLike.scala

```diff
@@ -0,0 +1,134 @@
+package com.mesosphere.usi.core.models
+
+import com.mesosphere.usi.core.models.resources.ResourceRequirement
+import com.typesafe.scalalogging.StrictLogging
+import org.apache.mesos.v1.{Protos => Mesos}
+
+import scala.collection.JavaConverters._
+
+sealed trait RunTemplateLike {
+  final lazy val allResources: Seq[ResourceRequirement] = {
+    val b = Seq.newBuilder[ResourceRequirement]
+    taskResourceRequirements.values.foreach { r =>
+      r.foreach(b += _)
+    }
+    executorResources.foreach { b += _ }
+    b.result()
+  }
+  val taskResourceRequirements: Map[PartialTaskId, Seq[ResourceRequirement]]
+  val executorResources: Seq[ResourceRequirement]
+  val role: String
+  private[usi] def buildOperation(
+      matchedOffer: Mesos.Offer,
+      taskIdStrategy: CurriedPodTaskIdStrategy,
+      executorResources: Seq[Mesos.Resource],
+      taskResources: Map[PartialTaskId, Seq[Mesos.Resource]])
+    : Either[Mesos.Offer.Operation.Launch, Mesos.Offer.Operation.LaunchGroup]
+}
+
+class LegacyLaunchBuilder(val role: String, taskBuilder: TaskBuilderLike) extends RunTemplateLike {
```

Comment:
This is a little confusing to me. Maybe it's the naming but why would a builder extend a template?

Reply:
How about making the build a field of the template?

Reply:
Yeah, WIP PR did not include attention to naming :) I've renamed it. I think this is an ok use of inheritance but am happy to hear your ideas on how composition would be better here.